### PR TITLE
raft: protect unstable.entries slice from appends

### PR DESCRIPTION
**Pull request by pavelkalinnikov**
This PR modifies `unstable.slice()` method to return a sub-slice using Go's full slice expression. This protects `unstable.entries` slice from appends that the callers of this method will be inclined to do (example one is in `unstable.truncateAndAppend()` method, which also got simpler with this PR).

This slice may ultimately bubble up to the client through `Ready` struct, which is another reason to protect it. Finally, the new code in `unstable.truncateAndAppend()` allocates once instead of twice.

### CODE REVIEW

**Review by nvanbenschoten — APPROVED** [review]
LGTM

File: log_unstable.go [hunk 1]
```diff
@@ -210,21 +210,27 @@ func (u *unstable) truncateAndAppend(ents []pb.Entry) {
 	default:
 		// truncate to after and copy to u.entries then append.
 		u.logger.Infof("truncate the unstable entries before index %d", after)
-		keep := u.slice(u.offset, after)
-		u.entries = append([]pb.Entry{}, keep...)
-		u.entries = append(u.entries, ents...)
+		u.entries = append(u.slice(u.offset, after), ents...)
```
[the remark this answers is not in the document]
Comment (pavelkalinnikov) [reply log_unstable.go]: Done.

File: log_unstable.go [hunk 2]
```diff
@@ -194,37 +194,43 @@ func (u *unstable) restore(s pb.Snapshot) {
...
+// TODO(pavelkalinnikov): this, and similar []pb.Entry slices, may bubble up all
+// the way to the application code through Ready struct. Protect other slices
+// similarly, and document how the client can use them.
 func (u *unstable) slice(lo uint64, hi uint64) []pb.Entry {
 	u.mustCheckOutOfBounds(lo, hi)
-	return u.entries[lo-u.offset : hi-u.offset]
+	// NB: use the full slice expression to limit what the caller can do with the
+	// returned slice. For example, an append will reallocate and copy this slice
+	// instead of corrupting the neighbouring u.entries.
+	return u.entries[lo-u.offset : hi-u.offset : hi-u.offset]
```
Comment (ahrtr) [inline log_unstable.go:233]: nice!

**Review by ahrtr — APPROVED** [review]
LGTM

Thank you @pavelkalinnikov

**Pull request merged**